pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -3402,6 +3488,454 @@ def SeismicActivityExceptionRule(model, k, t):
+                                model.p_lambda_Pipeline[key], to_units=pyunits.meter
+                            ),
+                            pyunits.convert(
+                                mh.p_diameter_Pipeline_eff[key], to_units=pyunits.meter
+                            ),
+                            pyunits.convert_value(
+                                value(model.v_F_Piped[key, t0]),
+                                from_units=pyunits.koil_bbl / pyunits.week,
+                                to_units=pyunits.m**3 / pyunits.s,
+                            ),
+                        )
+
+        def NodePressureRule(b, l1, l2, t1):
+            if value(model.v_F_Piped[l1, l2, t1]) > 0.01:
+                constraint = (
+                    b.v_Pressure[l1, t1] + model.p_Elevation[l1] * mh.p_rhog
+                    == b.v_Pressure[l2, t1]
+                    + model.p_Elevation[l2] * mh.p_rhog
+                    + b.p_HW_loss[l1, l2, t1] * mh.p_rhog
+                    - b.v_PumpHead[l1, l2, t1] * mh.p_rhog
+                    + b.v_ValveHead[l1, l2, t1] * mh.p_rhog
+                )
+                return process_constraint(constraint)
+            else:
+                return Constraint.Skip
+
+        mh.NodePressure = Constraint(
+            model.s_LLA,
+            model.s_T,
+            rule=NodePressureRule,
+            doc="Pressure at Node L",
+        )
+
+        def MAOPressureRule(b, l1, t1):
+            constraint = b.v_Pressure[l1, t1] <= b.p_Max_AOP
+            return process_constraint(constraint)
+
+        mh.MAOPressure = Constraint(
+            model.s_L - model.s_P,
+            model.s_T,
+            rule=MAOPressureRule,
+            doc="Max allowable pressure rule",
+        )
+
+        mh.objective = Objective(
+            expr=(
+                sum(
+                    sum(
+                        (mh.v_PumpHead[key, t] * 10 + mh.v_ValveHead[key, t])
+                        for key in model.s_LLA
+                    )
+                    for t in model.s_T
+                )
+            ),
+            sense=minimize,
+            doc="Objective function",
+        )
+
+        opt.solve(mh, tee=True)
+
+    elif model.config.hydraulics == Hydraulics.co_optimize:
+        """
+        For the co-optimize method, first, the strategic model is cloned and then,
+        constraints for computing pressures at every node using
+        Hazen-Williams equation is added to the model. The original objective
+        function of minmizing the cost is modified to include of the cost of
+        operating pumps and valves as a function of the deltaP.
+        """
+        # Clone the model to add hydraulics constraints
+        # The cloned model will be a MINLP formulation. The exisiting objective
+        # must be modified to add operating costs for pumps and valves.
+        mh = model.clone()
+        del mh.objective
+
+        mh.p_HW_material_factor_pipeline = Param(
```

Comment:
Beginning here, there's a good bit of code which is repeated from the `Hydraulics.post_process` section. Is there a way to consolidate/reformulate to avoid repeating a bunch of things?

Reply:
I agree, however, the trade off is either repeat the definition of variables or check for the hydraulics config arg multiple times. I have now switched back to checking the config arg multiple times and build the model appropriately

Reply:
Resolving.